### ISSUE TITLE
tf.keras.models.load_model does not load optimizer's weigths

### ISSUE THREAD

**Issue by alvaro-garcia-carrasco** [tensorflow/tensorflow#49940]
**System information**
Tensorflow: 2.5.0

**Describe current behaviour**
When using `tf.keras.models.load_model()` on a savedmodel containing a keras model saved using `tf.keras.Model.save(include_optimizer=True)`, the optimizer's weights are NOT loaded

**Describe expected behaviour**
The reconstructed model should contain the optimizer's weights to be able to resume training.

**Standalone code to reproduce issue**
https://colab.research.google.com/drive/1Y0y-pvE2QDw1Wj68XnbpLaUggXAXyzcx?usp=sharing

**Comment by UsharaniPagadala** [comment]
@alvaro-garcia-carrasco 

I saved model in hdf5 format and could be able to see the optimizer's weights loaded in the reconstructed model.Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/c270c6a1bc96365ac18ad4522e1b827b/load_savedmodel.ipynb).Thanks

**Comment by alvaro-garcia-carrasco** [comment]
Hi @UsharaniPagadala! Thanks for the response. Saving/loading the model as hdf5 format works fine and it is a potential workaround, but one should still be able to save/load it using the savedmodel, which is the recommended format. 

In the documentation of how to save/load keras models they provide a savedmodel example that does not work (it does not reconstruct the optimizer) https://www.tensorflow.org/guide/keras/save_and_serialize#savedmodel_format

**Comment by alvaro-garcia-carrasco** [comment]
It keeps the parameters of the optimizer such as learning rate and decay but I think that it does not restore the weights

**Comment by UsharaniPagadala** [comment]
@alvaro-garcia-carrasco 

The SavedModel contains `assets  saved_model.pb  variables
`

The model architecture, and training configuration (including the optimizer, losses, and metrics) are stored in saved_model.pb. The weights are saved in the variables/ directory.For more info refer [this](https://www.tensorflow.org/guide/keras/save_and_serialize).Thanks

**Comment by alvaro-garcia-carrasco** [comment]
@UsharaniPagadala 

The internals of how the graph and the variables are stored is not relevant. I am trying to load a SavedModel following the instructions in the documentation https://www.tensorflow.org/guide/keras/save_and_serialize#savedmodel_format and the optimizer's weights are not being loaded, as shown in the standalone code I provided to reproduce the issue.

**Comment by UsharaniPagadala** [comment]
@alvaro-garcia-carrasco 

`model.save_weights` method will help you to extract weights from saved model format .Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/d595f97fad666b3e7d6f96c6ed9e3c96/copy-of-load_savedmodel.ipynb).Thanks

**Comment by alvaro-garcia-carrasco** [comment]
@UsharaniPagadala This issue is not related on how to save the weights of a keras model. 

I refer again to mi initial bug description. The problem is that tf.keras.models.load_model does not load optimizer's weigths, as shown by the standalone code I provided.

**Comment by jvishnuvardhan** [comment]
@alvaro-garcia-carrasco By default, the `compile` argument under `load_model` is set to `True`. If you replace that with `False`, then you can see optimizer weights as you are expecting. Please see below for the change
Please change the following line 
`reconstructed_model = tf.keras.models.load_model("my_model",compile=True)`

Change the above line to 
`reconstructed_model = tf.keras.models.load_model("my_model",compile=False)
`
[Here](https://colab.research.google.com/gist/jvishnuvardhan/cff2f8fe46049754b24f36ea43fb1cbb/load_savedmodel.ipynb) for reference. In the gist, I have shown that the variables before saving the model and after loading the model are same. Thanks!

**Comment by sg10** [comment]
@UsharaniPagadala I am dealing with the same issue, but unfortunately there seems to be a misunderstanding.

- you are referring to the weights of the model itself (trainable variables as part of e.g. Dense layers, Conv layers, Batch Norm, etc.)
- the issue described by @alvaro-garcia-carrasco is related to the optimizer weights, these are not the same as the model weights.

The Adam optimizer (which is the default optimizer in Keras) contains **additional** weights (variables) that are necessary for the weight updates during training. These are crucial for continuing training from a checkpoint.

You can access the optimizer weights on any Keras model with `model.optimizer.get_weights`. These are not the same as `model.get_weights`. 

I created a colab to demonstrate the difference: https://colab.research.google.com/drive/1OaKlc9tYhsvs8BztpjpLEg6QD4uIIVSZ?usp=sharing

**Comment by alvaro-garcia-carrasco** [comment]
@jvishnuvardhan Thanks for the reply. The reason to restate the optimizer is to continue training. The keras model needs to be compiled after loading to continue training, which erases the optimizer weights again :-(. I feel we are getting closer though

**Comment by saumikn** [comment]
@jvishnuvardhan In your gist, the optimizer weights aren't the same - `model.optimizer.weights` has a length of `5`, while `reconstructed_model.optimizer.weights` has a length of `4`. In any case, I'm not sure that the bug with retraining models is because of the optimizer weights. I have a [gist](https://colab.research.google.com/drive/1MarUl9Q-Ny6BI5e7kRQLsJoxuqxv4GEr?usp=sharing) which shows that even if I overwrite the saved model weights with the original optimizer weights, the model doesn't retrain in the same way.

@alvaro-garcia-carrasco Were you able to get the training resumption to work? I've tried to hack this as you mentioned, but I wasn't able to figure out how to get this working correctly.

**Comment by jvishnuvardhan** [comment]
@alvaro-garcia-carrasco and @saumikn Thanks. Agree that this is a bug. We will work on it. If you want to contribute, please feel free to raise a PR to update the source code. Thanks!

**Comment by jvishnuvardhan** [comment]
Duplicate of #https://github.com/tensorflow/tensorflow/issues/44670

Let's track the progress in that issue. Thanks!

**Comment by sachinprasadhs** [comment]
please make sure you call load_weights() if you want to get the optimizer weights. For example:

```
def roundtrip(model: keras.Model) -> keras.Model:
    save_dir = "/tmp/mymodel"
    model.save(save_dir)
    restored = keras.models.load_model(save_dir)
    restored.load_weights(save_dir)
    return restored
```
